### PR TITLE
Add a db to store the different actions the bot made

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

**Comment by calixteman** [comment]
@La0, could you review the db file please:
https://github.com/mozilla/relman-auto-nag/pull/542/files#diff-f1b87de1e4a7e464afc84bab33adf5f4

**Comment by sylvestre** [comment]
```
    auto_nag/bzcleaner.py line 150 in get_summary
      return '...' if bug['groups'] else bug['summary']
   KeyError: 'groups'
```

and it would be nice to have some tests for that

File: auto_nag/nag_me.py [hunk 1]
```diff
@@ -160,6 +161,7 @@ def send_mails(self, title, dryrun=False):
                 login=login_info,
                 dryrun=dryrun,
             )
+            db.SentEmail.add(self.name(), receivers, 'individual')
```
Comment (La0) [inline auto_nag/nag_me.py]: Why not put this line in `mail.send` ?

File: auto_nag/db.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+from auto_nag.history import History
+
+
+Base = declarative_base()
+lock_path = utils.get_config('common', 'lock')
+db_path = utils.get_config('common', 'database')
+engine = create_engine(db_path)
+DBSession = sessionmaker(bind=engine)
+Base.metadata.bind = engine
+session = DBSession()
```
Comment (La0) [inline auto_nag/db.py]: That part seems really complex, there must be a helper to open a session somewhere

File: auto_nag/db.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+
+def clear():
```
Comment (La0) [inline auto_nag/db.py]: Do you need that ? it does not seem used

File: auto_nag/db.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+    Base.metadata.drop_all()
+    session.commit()
+
+
+def create():
```
Comment (La0) [inline auto_nag/db.py]: I recommend switching to Alembic for managing the db state

File: auto_nag/db.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+            date = dateutil.parser.parse(date)
+        date = int(calendar.timegm(date.timetuple()))
+        return date
+    if default == 'now':
+        return lmdutils.get_timestamp('now')
+    return default
+
+
+class Tool(Base):
+    __tablename__ = 'tools'
```
Comment (La0) [inline auto_nag/db.py]: I would add a prefix to the tables

File: auto_nag/db.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    name = Column(String(128), unique=True)
+    bugchanges = relationship('BugChange', backref='tool')
+    sentmails = relationship('SentEmail', backref='tool')
+
+    def __init__(self, name):
+        self.name = name
+
+    @staticmethod
+    def get(name):
```
Comment (La0) [inline auto_nag/db.py]: maybe rename to `get_or_create`, à la Django

File: auto_nag/db.py [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+        e = session.query(Tool).filter(Tool.name == name).first()
```
Comment (La0) [inline auto_nag/db.py]: I would rather use [one()](https://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.one) and create in the exception handler

File: auto_nag/db.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+                        BugChange.date >= start_date,
+                        BugChange.date < end_date,
+                    )
+                )
+            else:
+                rs = session.query(BugChange).filter(
+                    BugChange.date >= start_date, BugChange.date < end_date
+                )
+
+            return [r for r in rs]
```
Comment (La0) [inline auto_nag/db.py]: Why do you exec the query here ?

File: auto_nag/db.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+                    BugChange.date >= start_date,
+                    BugChange.date < end_date,
+                )
+
+            for r in res:
+                data[r.bugid] = True
+            return data
+
+    @staticmethod
+    def dump(path=''):
```
Comment (La0) [inline auto_nag/db.py]: A db structure change will make that really complex

Comment (La0) [reply auto_nag/db.py]: If you switch to alembic, you can make data migrations too (with forward/rollback)

File: auto_nag/db.py [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+        extra = self.extra.extra if self.extra else ''
+        return '<Bug change ({}): bug {}, the {}, extra={}>'.format(
+            self.tool, self.bugid, self.get_date(), extra
+        )
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class Email(Base):
```
Comment (La0) [inline auto_nag/db.py]: Maybe rename to user ?

File: auto_nag/db.py [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+            print(x)
+
+    def __repr__(self):
+        return self.extra
+
+    def __str__(self):
+        return self.__repr__()
+
+
+class SentEmail(Base):
```
Comment (La0) [inline auto_nag/db.py]: I would rename to Email, and add a state field (built/sent/failure)

File: auto_nag/db.py (hunk 5 above)
Comment (calixteman) [reply auto_nag/db.py]: like table_tools ?

Comment (calixteman) [reply auto_nag/db.py]: ok I prefixed them with autonag_

File: auto_nag/db.py (hunk 2 above)
Comment (calixteman) [reply auto_nag/db.py]: I didn't find anything for that, if you've any clues...

**Comment by sylvestre** [comment]
I am fine when @La0 is

File: auto_nag/db.py [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+Base = declarative_base()
+lock_path = utils.get_config('common', 'lock')
+db_url = utils.get_config('common', 'database')
+engine = create_engine(db_url)
+DBSession = sessionmaker(bind=engine)
+Base.metadata.bind = engine
+session = DBSession()
+
+
+def init():
```
Comment (La0) [inline auto_nag/db.py]: Is this still needed ? I don't find any ref to this function

File: auto_nag/db.py [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+                            'extra': extra,
+                        }
+                    )
+                json.dump(data, Out)
+        else:
+            for x in res:
+                print(x)
+
+    @staticmethod
+    def read_dict(data):
```
Comment (La0) [inline auto_nag/db.py]: Nit: rename to `import_from_dict` to show that you impact the DB

File: auto_nag/tests/test_db.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+
+import dateutil.parser
+import unittest
+
+import auto_nag.db as db
+
+
+class TestDB(unittest.TestCase):
+
+    HISTORY = [
```
Comment (La0) [inline auto_nag/tests/test_db.py]: Nit: move that to a json/yaml/whatever file ?

File: auto_nag/db.py (hunk 2 above)
Comment (La0) [reply auto_nag/db.py]: Nah you're right, it does not get [much simpler](https://docs.sqlalchemy.org/en/13/orm/session_basics.html)

**Comment by sylvestre** [comment]
@calixteman are we good to land that?

File: auto_nag/db.py (hunk 12 above)
Comment (calixteman) [reply auto_nag/db.py]: yep, just to init the db with data from bugzilla

**Comment by sylvestre** [comment]
Bastien is on pto today, so, I will go ahead and approve it

**Pull request merged**